### PR TITLE
[API Implementation]: Expose `System.Drawing.Graphics.NativeGraphics`.

### LINKED ISSUE #8833: API Proposal: Expose `System.Drawing.Graphics.NativeGraphics`.

### LINKED ISSUE #39406: API Proposal: Expose `System.Drawing.Graphics.NativeGraphics`.
## Background and Motivation

In Windows Forms painting it would be helpful if we had the option to plug our own P/Invokes calls into existing `Graphics` objects. We could, for example, have lighter-weight `GpPen` and `GpBrush` wrappers (structs perhaps) and invoke the GDI+ API directly with them via the exposed `Graphics` pointer.

While we could create our own `GpGraphics` object, that usually isn't practical as most of our public API works with `Graphics`.

Note that we already utilize the `IDeviceContext` API to extract the underlying `HDC`.

## Proposed API

```diff
namespace System.Drawing
{
    public class Graphics
    {
+    public IntPtr NativeGraphics { get; }
    }
}
```

## Risks

No specific risks outside of what comes normally with accessing backing handles. You can delete the backing object or try to use the handle after the `Graphics` closes the native handle. Not much different than what you'd be running into using a disposed `Graphics`.

Note that we already expose `IDeviceContext` which is much more risky as everything on `Graphics` throws when you're in-between `GetHDC()` and `ReleaseHDC()`. The throw there comes from GDI+ as it doesn't want to conflict with the usage of the backing HDC. GDI+ maintains and restores the state of the HDC outside of those calls.

### PR DESCRIPTION
**Pull request by deeprobin**
Proposal implementation of dotnet/winforms#8833 (closes dotnet/winforms#8833)

## Proposal 

```cs
namespace System.Drawing
{
    public partial class Graphics
    {
        public IntPtr Handle { get; }
        public static Graphics FromHandle(IntPtr handle);
    }
    public partial class Brush
    {
        public IntPtr Handle { get; }
        public static Brush FromHandle(IntPtr handle);
    }
    public partial class Pen
    {
        public IntPtr Handle { get; }
        public static Pen FromHandle(IntPtr handle);
    }
    public partial class Image
    {
        public IntPtr Handle { get; }
        public static Image FromHandle(IntPtr handle);
    }

    public partial class Region
    {
        public IntPtr Handle { get; }
        public static Region FromHandle(IntPtr handle);
    }
}
namespace System.Drawing.Drawing2D
{
    public partial class Matrix
    {
        public IntPtr Handle { get; }
        public static Matrix FromHandle(IntPtr handle);
    }

    public partial class GraphicsPath
    {
        public IntPtr Handle { get; }
        public static GraphicsPath FromHandle(IntPtr handle);
    }
}
```

## Current state of implementation
* [x] Implementation
* [x] Ref Assembly
* [x] Documentation comments
* [x] Fix project files
* [x] Tests

/cc @danmoseley

### CODE REVIEW

**Comment by danmoseley** [comment]
> I think because we forward in most cases the method to the private constructor, that we need no tests

We do (or should) test all public API even pass throughs - the code will exist for years, refactoring happens over time and can introduce some issue or another. Plus, it ekes out a little more code coverage when we scan for missing coverage.

**Comment by deeprobin** [comment]
@danmoseley Why is there still a build pipeline for NET Framework (which fails)? I thought that is officially not supported anymore?

**Comment by teo-tsirpanis** [comment]
@deeprobin, System.Drawing.Common tests are executed on both .NET Framework and modern .NET to ensure behavioral compatibility between the two editions of the same library.

**Comment by danmoseley** [comment]
@deeprobin how's this one going? do you need help?

You have removed some entries from the ref, which creates a breaking change for code building against the non-netcoreapp target framework, eg., for .NET Standard. Existing API should stay in there. New API can go in the ref/...netcoreapp.cs file. 

If possible, please use https://github.com/dotnet/runtime/blob/4679edc6be3eb91263a7acfe542dfd30b53ed2bf/docs/coding-guidelines/updating-ref-source.md#for-most-assemblies-within-libraries -- it doesn't understand multiple ref files, but you can copy/paste the result to try to get formatting like the tool would create.

**Comment by deeprobin** [comment]
> @deeprobin how's this one going? do you need help?
> 
> You have removed some entries from the ref, which creates a breaking change for code building against the non-netcoreapp target framework, eg., for .NET Standard. Existing API should stay in there. New API can go in the ref/...netcoreapp.cs file.
> 
> If possible, please use https://github.com/dotnet/runtime/blob/4679edc6be3eb91263a7acfe542dfd30b53ed2bf/docs/coding-guidelines/updating-ref-source.md#for-most-assemblies-within-libraries -- it doesn't understand multiple ref files, but you can copy/paste the result to try to get formatting like the tool would create.

Oh yes, I think I fixed that.

CI is now failing: `exit code 139 means SIGSEGV Illegal memory access. Deref invalid pointer, overrunning buffer, stack overflow etc. Core dumped.` (because the test-pointer does not exist)

How would you construct a test for this?

Creating 2 Handle-objects and swapping the `Handle` property? Or do you have a better idea to test this?

**Comment by danmoseley** [comment]
Not sure of your question. If you're asking - how would we construct a test that expected a fatal error like a segfault? - we do not test for fatal errors, except possibly in some exceptional cases (and then by using RemoteExecutor.Invoke).

**Comment by deeprobin** [comment]
@danmoseley I have looked at the tests again. The CI throws gdiplus errors. 
Do you think this is a case of https://github.com/dotnet/runtime/issues/22221.

<https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-63058-merge-ffd8398ff9224cd9a3/System.Drawing.Common.Tests/1/console.340b41fa.log?sv=2019-07-07&se=2022-01-29T01%3A02%3A29Z&sr=c&sp=rl&sig=2gsJHhJexH7K8z5Agh30IsXzPwEOXfezeP2vnLL0oNM%3D>

```
      System.TypeInitializationException : The type initializer for 'Gdip' threw an exception.
      ---- System.DllNotFoundException : Unable to load shared library 'libgdiplus' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: Error loading shared library liblibgdiplus: No such file or directory
      Stack Trace:
    System.Drawing.Drawing2D.Tests.MatrixTests.FromHandle [FAIL]
        /_/src/libraries/System.Drawing.Common/src/Microsoft.Interop.DllImportGenerator/Microsoft.Interop.DllImportGenerator/GeneratedDllImports.g.cs(518,0): at System.Drawing.SafeNativeMethods.Gdip.GdipCreateMatrix(IntPtr& matrix)
        /_/src/libraries/System.Drawing.Common/src/System/Drawing/Drawing2D/Matrix.cs(18,0): at System.Drawing.Drawing2D.Matrix..ctor()
        /_/src/libraries/System.Drawing.Common/tests/Drawing2D/MatrixTests.Core.cs(47,0): at System.Drawing.Drawing2D.Tests.MatrixTests.FromHandle()
        ----- Inner Stack Trace -----
           at System.Drawing.SafeNativeMethods.Gdip.GdiplusStartup(IntPtr& token, StartupInput& input, StartupOutput& output)
        /_/src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs(33,0): at System.Drawing.SafeNativeMethods.Gdip..cctor()
```

**Comment by deeprobin** [comment]
@qmfrederik @eerhardt Can you perhaps give some thoughts on this, since you were involved in #22221?

**Comment by deeprobin** [comment]
I have already implemented the change with the ActiveIssueAttribute. In case of doubt we can also revert this.

https://github.com/deeprobin/runtime/tree/f39d2a3aeef9da389e9de5da7c77ad798e6461f1

**Comment by danmoseley** [comment]
@deeprobin your tests are causing a double-dispose, eg:
```c#
        public void FromHandle()
        {
            var region = new Region();

            var expectedHandle = region.Handle;
            var actualRegion = Region.FromHandle(expectedHandle);

            IntPtr actualHandle = actualRegion.Handle;
            Assert.Equal(expectedHandle, actualHandle);

            // Do not dispose `region` because it is the same handle
            actualRegion.Dispose();
        }
```
Here you're disposing `actualRegion` which is calling [GdipDeleteRegion](https://github.com/danmoseley/runtime/blob/983b193c1e509e5b8ed7f2fb0079b47696f256d7/src/libraries/System.Drawing.Common/src/System/Drawing/Region.cs#L102) with this handle. But the Region object is IDisposable - so when it becomes unreachable and is collected, the GC will put it on the finalizer queue. And the finalizer will call dispose, which will call GdipDeleteRegion on the same handle. You see in the log that GDI+ is returning 4, which means [ObjectBusy](https://docs.microsoft.com/en-us/windows/win32/api/Gdiplustypes/ne-gdiplustypes-status). I am guessing that GdipDeleteRegion may be safe to call twice on the same handle, but not concurrently, which is likely what's happening here.

The fix is likely to change `var region = new Region();` to `using var region = new Region();`. This will cause the Dispose to happen deterministically. The same issue applies to the other tests.

Because you've got 8 PR's active at this time, I'm going to close this one. Please reactivate when you believe all the tests should be passing again. Thanks.

File: src/libraries/System.Drawing.Common/ref/System.Drawing.Common.netcoreapp.cs [hunk 1]
```diff
@@ -19,13 +19,46 @@ public sealed partial class Graphics
         public void GetContextInfo(out PointF offset) { throw null; }
         [System.Runtime.Versioning.SupportedOSPlatformAttribute("windows")]
         public void GetContextInfo(out PointF offset, out Region? clip) { throw null; }
+        public System.IntPtr Handle { get { throw null; } }
+        public static System.Drawing.Graphics FromHandle(System.IntPtr handle) { throw null; }
+    }
+
```
Comment (danmoseley) [inline src/libraries/System.Drawing.Common/ref/System.Drawing.Common.netcoreapp.cs:25]: Our goal is to either generate these files (as discussed earlier) or make it look like the generator created it. The generator won't insert blank lines like this as we would in normal code.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Unix.cs [hunk 2]
```diff
@@ -901,5 +901,19 @@ public void Widen(Pen pen, Matrix? matrix, float flatness)
             int s = Gdip.GdipWidenPath(_nativePath, pen.NativePen, m, flatness);
             Gdip.CheckStatus(s);
         }
+
+        /// <summary>
+        /// The underlying graphics path handle
+        /// </summary>
+        public IntPtr Handle => _nativePath;
```
Comment (danmoseley) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Unix.cs:908]: We try to order fields then properties then methods, with public before private. I'm surprised the editorconfig isn't flagging this for you.

**Pull request closed without merging**